File: src/main/java/no/fintlabs/dispatch/DispatchService.java
package no.fintlabs.dispatch;

import lombok.extern.slf4j.Slf4j;
import no.fintlabs.dispatch.converting.InstanceFlowHeadersToInstanceHeadersEntityConvertingService;
import no.fintlabs.dispatch.converting.InstanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService;
import no.fintlabs.dispatch.model.InstanceHeadersEntity;
import no.fintlabs.dispatch.model.InstanceReceiptDispatchEntity;
import no.fintlabs.dispatch.repository.InstanceHeadersRepository;
import no.fintlabs.dispatch.repository.InstanceReceiptDispatchRepository;
import no.fintlabs.flyt.kafka.headers.InstanceFlowHeaders;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import reactor.core.publisher.Mono;
import reactor.core.publisher.Sinks;
import reactor.core.scheduler.Schedulers;

import java.util.List;
import java.util.Optional;

@Service
@Slf4j
public class DispatchService {

    public static final int EGRUNNERVERV_SOURCE_APPLICATION_ID = 2;
    private final WebClientRequestService webClientRequestService;
    private final InstanceFlowHeadersToInstanceHeadersEntityConvertingService
            instanceFlowHeadersToInstanceHeadersEntityConvertingService;
    private final InstanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService
            instanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService;
    private final InstanceHeadersRepository instanceHeadersRepository;
    private final InstanceReceiptDispatchRepository instanceReceiptDispatchRepository;
    private final Sinks.Many<InstanceHeadersEntity> newInstanceHeadersEntitySink;

    public DispatchService(
            WebClientRequestService webClientRequestService,
            InstanceFlowHeadersToInstanceHeadersEntityConvertingService instanceFlowHeadersToInstanceHeadersEntityConvertingService,
            InstanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService instanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService,
            InstanceHeadersRepository instanceHeadersRepository,
            InstanceReceiptDispatchRepository instanceReceiptDispatchRepository
    ) {
        this.webClientRequestService = webClientRequestService;
        this.instanceFlowHeadersToInstanceHeadersEntityConvertingService =
                instanceFlowHeadersToInstanceHeadersEntityConvertingService;
        this.instanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService =
                instanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService;
        this.instanceHeadersRepository = instanceHeadersRepository;
        this.instanceReceiptDispatchRepository = instanceReceiptDispatchRepository;
        newInstanceHeadersEntitySink = Sinks.many().unicast().onBackpressureBuffer();
        newInstanceHeadersEntitySink.asFlux().subscribe(this::dispatch);
    }


    public void handleNewInstance(InstanceFlowHeaders instanceFlowHeaders) {
        if (instanceFlowHeaders.getSourceApplicationId() != EGRUNNERVERV_SOURCE_APPLICATION_ID) {
            return;
        }

        InstanceHeadersEntity instanceHeadersEntity = saveInstanceHeaders(instanceFlowHeaders);
        newInstanceHeadersEntitySink.tryEmitNext(instanceHeadersEntity);
    }

    public synchronized void dispatch(InstanceHeadersEntity instanceHeadersEntity) {
        convertAndTransferToInstanceReceiptDispatch(instanceHeadersEntity)
                .ifPresent(this::dispatchInstanceReceipt);
    }

    @Scheduled(
            initialDelayString = "${fint.flyt.egrunnerverv.instance-dispatch-initial-delay}",
            fixedDelayString = "${fint.flyt.egrunnerverv.instance-dispatch-fixed-delay}"
    )
    private synchronized void dispatchAll() {
        List<InstanceHeadersEntity> instanceHeaders = instanceHeadersRepository.findAll();
        if (!instanceHeaders.isEmpty()) {
            log.info("Converting and transferring {} instance header entities to instance receipt dispatch entities", instanceHeaders.size());
            instanceHeaders.forEach(
                    this::convertAndTransferToInstanceReceiptDispatch
            );
        }

        List<InstanceReceiptDispatchEntity> instanceReceipts = instanceReceiptDispatchRepository.findAll();
        if (!instanceReceipts.isEmpty()) {
            log.info("Dispatching {} instance receipt dispatch entities", instanceReceipts.size());
            instanceReceipts.forEach(
                    this::dispatchInstanceReceipt
            );
        }
    }

    private InstanceHeadersEntity saveInstanceHeaders(InstanceFlowHeaders instanceFlowHeaders) {
        InstanceHeadersEntity instanceHeadersEntity = instanceFlowHeadersToInstanceHeadersEntityConvertingService
                .convert(instanceFlowHeaders);
        log.info(
                "Saving InstanceHeadersEntity for sourceApplicationInstanceId={}",
                instanceFlowHeaders.getSourceApplicationInstanceId()
        );
        instanceHeadersRepository.save(instanceHeadersEntity);
        return instanceHeadersEntity;
    }

    private Optional<InstanceReceiptDispatchEntity> convertAndTransferToInstanceReceiptDispatch(
            InstanceHeadersEntity instanceHeadersEntity
    ) {

        String sourceApplicationInstanceId = instanceHeadersEntity.getSourceApplicationInstanceId();

        log.info(
                "Converting InstanceHeadersEntity to InstanceReceiptDispatchEntity for sourceApplicationInstanceId={}",
                sourceApplicationInstanceId
        );

        try {
            Optional<InstanceReceiptDispatchEntity> instanceReceiptDispatchEntity =
                    instanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService.convert(instanceHeadersEntity)
                            .map(entity -> {
                                log.info(
                                        "Saving InstanceReceiptDispatchEntity for sourceApplicationInstanceId={}",
                                        sourceApplicationInstanceId
                                );
                                return instanceReceiptDispatchRepository.save(entity);
                            });

            if (instanceReceiptDispatchEntity.isPresent()) {
                log.info(
                        "Deleting InstanceHeadersEntity for sourceApplicationInstanceId={}",
                        sourceApplicationInstanceId
                );
                instanceHeadersRepository.delete(instanceHeadersEntity);
                log.info(
                        "Successfully converted and transferred InstanceHeadersEntity to" +
                                " InstanceReceiptDispatchEntity for sourceApplicationInstanceId={}",
                        sourceApplicationInstanceId
                );
            }

            return instanceReceiptDispatchEntity;
        } catch (RuntimeException e) {
            log.error("Converting and transferring of InstanceHeadersEntity to InstanceReceiptDispatchEntity failed" +
                    " for sourceApplicationInstanceId={}", sourceApplicationInstanceId, e);
        }

        return Optional.empty();
    }

    private void dispatchInstanceReceipt(InstanceReceiptDispatchEntity entity) {
        final String id = entity.getSourceApplicationInstanceId();

        webClientRequestService
                .dispatchInstance(entity)
                .flatMap(ok ->
                        Mono.fromRunnable(() -> instanceReceiptDispatchRepository.delete(ok))
                                .subscribeOn(Schedulers.boundedElastic())
                )
                .onErrorResume(WebClientResponseException.class, wre -> {
                    int code = wre.getRawStatusCode();
                    boolean terminal = (code >= 400 && code < 500 && code != 429);
                    if (terminal) {
                        return Mono.fromRunnable(() -> {
                                    log.warn("Terminal {} for {}, deleting row.", code, id);
                                    instanceReceiptDispatchRepository.deleteById(id);
                                })
                                .subscribeOn(Schedulers.boundedElastic())
                                .then(Mono.empty());
                    }
                    log.warn("Transient {} for {}, will retry later.", code, id);
                    return Mono.error(wre);
                })
                .onErrorComplete()
                .block();
    }
}

File: src/main/kotlin/no/novari/flyt/egrunnerverv/gateway/infrastructure/dispatch/DispatchService.kt
package no.novari.flyt.egrunnerverv.gateway.infrastructure.dispatch

import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.annotation.PreDestroy
import net.logstash.logback.argument.StructuredArguments.kv
import no.novari.flyt.egrunnerverv.gateway.infrastructure.dispatch.converting.InstanceFlowHeadersToInstanceHeadersEntityConvertingService
import no.novari.flyt.egrunnerverv.gateway.infrastructure.dispatch.converting.InstanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService
import no.novari.flyt.egrunnerverv.gateway.infrastructure.dispatch.model.InstanceHeadersEntity
import no.novari.flyt.egrunnerverv.gateway.infrastructure.dispatch.model.InstanceReceiptDispatchEntity
import no.novari.flyt.egrunnerverv.gateway.infrastructure.dispatch.repository.InstanceHeadersRepository
import no.novari.flyt.egrunnerverv.gateway.infrastructure.dispatch.repository.InstanceReceiptDispatchRepository
import no.novari.flyt.kafka.instanceflow.headers.InstanceFlowHeaders
import org.springframework.http.HttpStatus
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClientResponseException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

@Service
class DispatchService(
    private val restClientRequestService: RestClientRequestService,
    private val instanceFlowHeadersToInstanceHeadersEntityConvertingService:
        InstanceFlowHeadersToInstanceHeadersEntityConvertingService,
    private val instanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService:
        InstanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService,
    private val instanceHeadersRepository: InstanceHeadersRepository,
    private val instanceReceiptDispatchRepository: InstanceReceiptDispatchRepository,
) {
    private val dispatchExecutor: ExecutorService = Executors.newSingleThreadExecutor()

    fun handleNewInstance(instanceFlowHeaders: InstanceFlowHeaders) {
        log.atDebug {
            message = "Received instance flow headers"
            arguments =
                arrayOf(
                    kv("sourceApplicationId", instanceFlowHeaders.sourceApplicationId),
                    kv("sourceApplicationInstanceId", instanceFlowHeaders.sourceApplicationInstanceId),
                    kv("sourceApplicationIntegrationId", instanceFlowHeaders.sourceApplicationIntegrationId),
                    kv("archiveInstanceId", instanceFlowHeaders.archiveInstanceId),
                )
        }

        if (instanceFlowHeaders.sourceApplicationId != EGRUNNERVERV_SOURCE_APPLICATION_ID) {
            log.atDebug {
                message = "Skipping instance flow headers from another source application"
                arguments =
                    arrayOf(
                        kv("sourceApplicationId", instanceFlowHeaders.sourceApplicationId),
                        kv("expectedSourceApplicationId", EGRUNNERVERV_SOURCE_APPLICATION_ID),
                    )
            }
            return
        }

        val instanceHeadersEntity = saveInstanceHeaders(instanceFlowHeaders)
        log.atDebug {
            message = "Submitting instance for asynchronous dispatch"
            arguments = instanceHeadersArguments(instanceHeadersEntity)
        }
        dispatchExecutor.submit { dispatch(instanceHeadersEntity) }
    }

    @Synchronized
    fun dispatch(instanceHeadersEntity: InstanceHeadersEntity) {
        log.atDebug {
            message = "Starting dispatch of InstanceHeadersEntity"
            arguments = instanceHeadersArguments(instanceHeadersEntity)
        }

        val instanceReceiptDispatchEntity = convertAndTransferToInstanceReceiptDispatch(instanceHeadersEntity)
        if (instanceReceiptDispatchEntity == null) {
            log.atDebug {
                message = "No InstanceReceiptDispatchEntity produced, dispatch is skipped"
                arguments = instanceHeadersArguments(instanceHeadersEntity)
            }
            return
        }

        dispatchInstanceReceipt(instanceReceiptDispatchEntity)

        log.atDebug {
            message = "Finished dispatch attempt for InstanceHeadersEntity"
            arguments = instanceReceiptDispatchArguments(instanceReceiptDispatchEntity)
        }
    }

    @Scheduled(
        initialDelayString = "\${novari.flyt.egrunnerverv.dispatch.instance-initial-delay}",
        fixedDelayString = "\${novari.flyt.egrunnerverv.dispatch.instance-fixed-delay}",
    )
    @Synchronized
    fun dispatchAll() {
        val instanceHeaders = instanceHeadersRepository.findAll()
        if (instanceHeaders.isNotEmpty()) {
            log.atInfo {
                message = "Converting and transferring instance header entities to instance receipt dispatch entities"
                arguments = arrayOf(kv("size", instanceHeaders.size))
            }
            instanceHeaders.forEach(::convertAndTransferToInstanceReceiptDispatch)
        } else {
            log.atDebug {
                message = "No InstanceHeadersEntity rows found for scheduled dispatch"
            }
        }

        val instanceReceipts = instanceReceiptDispatchRepository.findAll()
        if (instanceReceipts.isNotEmpty()) {
            log.atInfo {
                message = "Dispatching instance receipt dispatch entities"
                arguments = arrayOf(kv("size", instanceReceipts.size))
            }
            instanceReceipts.forEach(::dispatchInstanceReceipt)
        } else {
            log.atDebug {
                message = "No InstanceReceiptDispatchEntity rows found for scheduled dispatch"
            }
        }
    }

    @PreDestroy
    fun shutdownExecutor() {
        log.atDebug {
            message = "Shutting down dispatch executor"
        }
        dispatchExecutor.shutdown()
    }

    private fun saveInstanceHeaders(instanceFlowHeaders: InstanceFlowHeaders): InstanceHeadersEntity {
        log.atDebug {
            message = "Converting InstanceFlowHeaders to InstanceHeadersEntity"
            arguments =
                arrayOf(
                    kv("sourceApplicationId", instanceFlowHeaders.sourceApplicationId),
                    kv("sourceApplicationInstanceId", instanceFlowHeaders.sourceApplicationInstanceId),
                    kv("sourceApplicationIntegrationId", instanceFlowHeaders.sourceApplicationIntegrationId),
                    kv("archiveInstanceId", instanceFlowHeaders.archiveInstanceId),
                )
        }

        val instanceHeadersEntity =
            instanceFlowHeadersToInstanceHeadersEntityConvertingService.convert(
                instanceFlowHeaders,
            )
        log.atInfo {
            message = "Saving InstanceHeadersEntity"
            arguments = arrayOf(kv("sourceApplicationInstanceId", instanceFlowHeaders.sourceApplicationInstanceId))
        }
        val savedInstanceHeadersEntity = instanceHeadersRepository.save(instanceHeadersEntity)

        log.atDebug {
            message = "Saved InstanceHeadersEntity"
            arguments = instanceHeadersArguments(savedInstanceHeadersEntity)
        }

        return savedInstanceHeadersEntity
    }

    private fun convertAndTransferToInstanceReceiptDispatch(
        instanceHeadersEntity: InstanceHeadersEntity,
    ): InstanceReceiptDispatchEntity? {
        val sourceApplicationInstanceId = instanceHeadersEntity.sourceApplicationInstanceId

        log.atInfo {
            message = "Converting InstanceHeadersEntity to InstanceReceiptDispatchEntity"
            arguments = arrayOf(kv("sourceApplicationInstanceId", sourceApplicationInstanceId))
        }
        log.atDebug {
            message = "Converting InstanceHeadersEntity to InstanceReceiptDispatchEntity"
            arguments = instanceHeadersArguments(instanceHeadersEntity)
        }

        return try {
            val instanceReceiptDispatchEntity =
                instanceHeadersEntityToInstanceReceiptDispatchEntityConvertingService
                    .convert(
                        instanceHeadersEntity,
                    )?.let { entity ->
                        log.atDebug {
                            message = "Converted InstanceHeadersEntity to InstanceReceiptDispatchEntity"
                            arguments = instanceReceiptDispatchArguments(entity)
                        }
                        log.atInfo {
                            message = "Saving InstanceReceiptDispatchEntity"
                            arguments = arrayOf(kv("sourceApplicationInstanceId", sourceApplicationInstanceId))
                        }
                        val savedEntity = instanceReceiptDispatchRepository.save(entity)

                        log.atDebug {
                            message = "Saved InstanceReceiptDispatchEntity"
                            arguments = instanceReceiptDispatchArguments(savedEntity)
                        }

                        savedEntity
                    }

            if (instanceReceiptDispatchEntity != null) {
                log.atInfo {
                    message = "Deleting InstanceHeadersEntity"
                    arguments = arrayOf(kv("sourceApplicationInstanceId", sourceApplicationInstanceId))
                }
                instanceHeadersRepository.delete(instanceHeadersEntity)
                log.atDebug {
                    message = "Deleted InstanceHeadersEntity after transfer to dispatch queue"
                    arguments = instanceHeadersArguments(instanceHeadersEntity)
                }
                log.atInfo {
                    message =
                        "Successfully converted and transferred InstanceHeadersEntity to InstanceReceiptDispatchEntity"
                    arguments = arrayOf(kv("sourceApplicationInstanceId", sourceApplicationInstanceId))
                }
            } else {
                log.atDebug {
                    message = "InstanceHeadersEntity conversion returned no dispatch entity"
                    arguments = instanceHeadersArguments(instanceHeadersEntity)
                }
            }

            instanceReceiptDispatchEntity
        } catch (exception: RuntimeException) {
            log.atError {
                cause = exception
                message = "Converting and transferring of InstanceHeadersEntity to InstanceReceiptDispatchEntity failed"
                arguments = arrayOf(kv("sourceApplicationInstanceId", sourceApplicationInstanceId))
            }
            null
        }
    }

    private fun dispatchInstanceReceipt(entity: InstanceReceiptDispatchEntity) {
        val id = entity.sourceApplicationInstanceId

        try {
            log.atDebug {
                message = "Dispatching InstanceReceiptDispatchEntity to ServiceNow"
                arguments = instanceReceiptDispatchArguments(entity)
            }
            restClientRequestService.dispatchInstance(entity)
            log.atDebug {
                message = "ServiceNow dispatch completed, deleting InstanceReceiptDispatchEntity"
                arguments = instanceReceiptDispatchArguments(entity)
            }
            instanceReceiptDispatchRepository.delete(entity)
            log.atDebug {
                message = "Deleted InstanceReceiptDispatchEntity after successful ServiceNow dispatch"
                arguments = instanceReceiptDispatchArguments(entity)
            }
        } catch (exception: RestClientResponseException) {
            val statusCode = exception.statusCode
            val retryable = statusCode == HttpStatus.TOO_MANY_REQUESTS || statusCode.is5xxServerError
            log.atDebug {
                cause = exception
                message = "ServiceNow dispatch failed with response"
                arguments =
                    arrayOf(
                        kv("statusCode", statusCode.value()),
                        kv("id", id),
                        kv("uri", entity.uri),
                        kv("responseBody", exception.responseBodyAsString),
                    )
            }
            if (!retryable) {
                log.atWarn {
                    cause = exception
                    message = "Terminal failure, deleting row."
                    arguments = arrayOf(kv("statusCode", statusCode.value()), kv("id", id), kv("uri", entity.uri))
                }
                instanceReceiptDispatchRepository.deleteById(id)
                log.atDebug {
                    message = "Deleted terminally failed InstanceReceiptDispatchEntity"
                    arguments =
                        arrayOf(
                            kv("statusCode", statusCode.value()),
                            kv("id", id),
                            kv("uri", entity.uri),
                        )
                }
            } else {
                log.atWarn {
                    cause = exception
                    message = "Transient failure, will retry later."
                    arguments = arrayOf(kv("statusCode", statusCode.value()), kv("id", id), kv("uri", entity.uri))
                }
            }
        } catch (exception: Exception) {
            log.atDebug {
                cause = exception
                message = "Unexpected failure during ServiceNow dispatch"
                arguments = instanceReceiptDispatchArguments(entity)
            }
            log.atWarn {
                cause = exception
                message = "Dispatch failed, will retry later."
                arguments = arrayOf(kv("id", id), kv("uri", entity.uri))
            }
        }
    }

    private fun instanceHeadersArguments(entity: InstanceHeadersEntity): Array<Any?> =
        arrayOf<Any?>(
            kv("sourceApplicationInstanceId", entity.sourceApplicationInstanceId),
            kv("sourceApplicationIntegrationId", entity.sourceApplicationIntegrationId),
            kv("archiveInstanceId", entity.archiveInstanceId),
        )

    private fun instanceReceiptDispatchArguments(entity: InstanceReceiptDispatchEntity): Array<Any?> =
        arrayOf<Any?>(
            kv("sourceApplicationInstanceId", entity.sourceApplicationInstanceId),
            kv("uri", entity.uri),
            kv("instanceReceipt", entity.instanceReceipt),
        )

    private companion object {
        private const val EGRUNNERVERV_SOURCE_APPLICATION_ID = 2L
        private val log = KotlinLogging.logger {}
    }
}
